README.md
Snzip, a compression/decompression tool based on snappy.
========================================================

What is snzip.
--------------

Snzip is one of command line tools using [snappy][]. This supports several file
formats; [framing-format][], [old framing-format][], [hadoop-snappy format][], [raw format][]
and obsolete three formats used by snzip, [snappy-java][] and [snappy-in-java][]
before official framing-format was defined. The default format is [framing-format][].

Notable Changes
---------------

The default format was changed to [framing-format][] in 1.0.0.
Set `--with-default-format=snzip` as a configure option to use obsolete snzip
format as the default format as before.

Installation
------------

### Install from a tar-ball

Download snzip-1.0.5.tar.gz from https://github.com/kubo/snzip/releases,
uncompress and untar it, and run configure.

    tar xvfz snzip-1.0.5.tar.gz
    cd snzip-1.0.5
    ./configure
    make
    make install

If you didn't install snappy under `/usr` or `/usr/local`, you need to specify
the location by `--with-snappy` as follows.

    # install snzip
    tar xvfz snzip-1.0.5.tar.gz
    cd snzip-1.0.5
    ./configure --with-snappy=/xxx/yyy/
    make
    make install

When both dynamic and static snappy libraries are available, the former
is used by default. The compiled `snzip` depends on `libsnappy.so`.
When `--with-static-snappy` is passed as a configure option, the latter
is used. The compiled `snzip` includes snappy library.

Note: `--with-static-snappy` isn't available on some platforms.

You can use `--with-default-format` to change the default compression format.

    ./configure --with-default-format=snzip

### Install as a rpm package

We don't provide rpm packages. You need to download snzip-1.0.5.tar.gz
from https://github.com/kubo/snzip/releases, create a rpm package as follows and
install it.

    # The rpm package will be created under $HOME/rpmbuild/RPMS.
    rpmbuild -tb snzip-1.0.5.tar.gz 

### Install from the latest source

To use source code in the github repository.

    git clone git://github.com/kubo/snzip.git
    cd snzip
    ./autogen.sh
    ./configure
    make
    make install

### Install a Windows package.

Download `snzip-1.0.5-win32.zip` or `snzip-1.0.5-win64.zip` from
https://github.com/kubo/snzip/releases and copy `snzip.exe` and `snunzip.exe`
to a directory in the PATH environment variable.

Usage
-----

### To compress file.tar:

    snzip file.tar

Compressed file name is `file.tar.sz` and the original file is deleted.
The file attributes such as timestamp, mode and permissions are not changed
as possible as it can.

The compressed file's format is [framing-format][]. You need to add an option `-t snappy-java` or
`-t snappy-in-java` to use other formats.

    snzip -t snappy-java file.tar

or

    snzip -t snappy-in-java file.tar

### To compress file.tar and output to standard out.

    snzip -c file.tar > file.tar.sz

or

    cat file.tar | snzip > file.tar.sz

You need to add an option `-t [format-name]` to use formats except [framing-format][].

### To create a new tar file and compress it.

    tar cf - files-to-be-archived | snzip > archive.tar.sz

### To uncompress file.tar.sz:

    snzip -d file.tar.sz

or

    snunzip file.tar.sz

Uncompressed file name is `file.tar` and the original file is deleted.
The file attributes such as timestamp, mode and permissions are not changed
as possible as it can.

If the program name includes `un` such as `snunzip`, it acts as `-d` is set.

The file format is automatically determined from the file header.
However it doesn't work for some file formats such as raw and Apple iWork .iwa.

### To uncompress file.tar.sz and output to standard out.

    snzip -dc file.tar.sz > file.tar
    snunzip -c file.tar.sz > file.tar
    snzcat file.tar.sz > file.tar
    cat file.tar.sz | snzcat > file.tar

If the program name includes `cat` such as snzcat, it acts as `-dc` is set.

### To uncompress a tar file and extract it.

    snzip -dc archive.tar.sz | tar xf -

Raw format
----------

Raw format is native format of snappy.
Unlike other formats, there are a few limitations:
(1) The total data length before compression must be known on compression.
(2) Automatic file format detection doesn't work on uncompression.
(3) The raw format support is enabled only when snzip is compiled for snappy 1.1.3 or upper.

### To compress file.tar as raw format:

    snzip -t raw file.tar

or

    snzip -t raw < file.tar > file.tar.raw

In these examples, snzip uses a file descriptor, which directly opens
the `file.tar` file, and gets the file length to be compressed.
However the following command doesn't work.

    cat file.tar | snzip -t raw > file.tar.raw

It uses a pipe. snzip cannot get the total length before compression.
The total length must be specified by the `-s` option in this case.

    cat file.tar | snzip -t raw -s "size of file.tar" > file.tar.raw

### To uncompress file.tar.sz compressed as raw format

    snzip -t raw -d file.tar.sz

or

    snunzip -t raw file.tar.sz

You need to set the `-t raw` option to tell snzip the format of the
file to be uncompressed.

Hadoop-snappy format
--------------------

Hadoop-snappy format is one of the compression formats used in Hadoop.
It uses its own framing format as follows:

* A compressed file consists of one or more blocks.
* A block consists of uncompressed length (big endian 4 byte integer) and one or more subblocks.
* A subblock consists of compressed length (big endian 4 byte integer) and raw compressed data.

### To compress a file as hadoop-snappy format

    snzip -t hadoop-snappy file_name

The default block size used by `snzip` for hadoop-snappy format is 256k.
It is same with the default value of the `io.compression.codec.snappy.buffersize`
parameter. If the block size used by `snzip` is larger than the parameter,
you would get an InternalError `Could not decompress data. Buffer length is too small`
while hadoop is reading a file compressed by snzip. You need to change the block
size by the `-b` option as follows if you get the error.

    # if  io.compression.codec.snappy.buffersize is 32768
    snzip -t hadoop-snappy -b 32768 file_name_to_be_compressed

### To uncompress a file compressed as haddoop-snappy format

    snzip -d compressed_file.snappy

The file format is guessed by the first 8 bytes of the file.

Apple iWork .iwa format
-----------------------

Apple iWork .iwa format is a file format used by Apple iWork. The format was
demystified [here](https://github.com/obriensp/iWorkFileFormat).
Basically the .iwa format consists of a Protobuf stream [compressed by Snappy](https://github.com/obriensp/iWorkFileFormat/blob/master/Docs/index.md#snappy-compression).

Snzip uncompresses .iwa files to Protbuf streams and compresses Protobuf streams
to .iwa files. You need to set `-t iwa` on compression and uncompression to
specify the file format.

SNZ File format
---------------

Note: This is obsolete format. The default format was changed to [framing-format].

The first three bytes are magic characters 'SNZ'.

The fourth byte is the file format version. It is 0x01.

The fifth byte is the order of the block size. The input data
is divided into fixed-length blocks and each block is compressed
by snappy. When it is 16 (default value), the block size is 16th
power of 2; 64 kilobytes.

The rest is pairs of a compressed data length and a compressed data block
The compressed data length is encoded as `snappy::Varint::Encode32()` does.
If the length is zero, it is the end of data.

Though the rest after the end of data is ignored for now, they
may be continuously read as a next compressed file as gzip does.

Note that the uncompressed length of each compressed data block must be
less than or equal to the block size specified by the fifth byte.

License
-------

2-clause BSD-style license.

[snappy]: https://github.com/google/snappy/blob/master/docs/README.md
[framing-format]: https://github.com/google/snappy/blob/master/framing_format.txt
[old framing-format]: https://github.com/google/snappy/blob/0755c815197dacc77d8971ae917c86d7aa96bf8e/framing_format.txt
[snappy-java]: https://github.com/xerial/snappy-java
[snappy-in-java]: https://github.com/dain/snappy
[raw format]: https://github.com/kubo/snzip#raw-format
[Hadoop-snappy format]: https://github.com/kubo/snzip#hadoop-snappy-format


NEWS
snzip 1.0.5 - 2021-12-06

* Fix a problem to read concatenated framing format data from stdin. (#28)
* Install snunzip and snzcat also by `make install` (#21)
* Add CMakeLists.txt to compile Windows binary.
* Change uncompressed file's file name when its suffix is not expected.

snzip 1.0.4 - 2016-10-02

* Add Apple iWork .iwa format support. (#14)

snzip 1.0.3 - 2016-03-12

* Add hadoop-snappy format support.
* Add raw format support.

snzip 1.0.2 - 2015-01-04

* Use SSE4.2 to calculate CRC32C if it is available.
* Add a getopt function and a Visual Studio 2013 project file for Windows.

snzip 1.0.1 - 2013-11-30

* Add a rpm spec file.
  (Contributed by Xianglei.)

snzip 1.0.0 - 2013-11-23

* The default format is changed to framing format.
* Change stdout behaviour like bzip2 or gzip.
  (Contribuetd by FUKUDA Masahiro.)

snzip 0.9.0 - 2012-08-05

snzip 0.0.2 - 2011-10-09

snzip 0.0.1 - 2011-10-08

* Initial experimental release


snzip-format.c
/* -*- indent-tabs-mode: nil -*-
 *
 * Copyright 2011-2012 Kubo Takehiro <kubo@jiubao.org>
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *    1. Redistributions of source code must retain the above copyright notice, this list of
 *       conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright notice, this list
 *       of conditions and the following disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS ''AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are those of the
 * authors and should not be interpreted as representing official policies, either expressed
 * or implied, of the authors.
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <snappy-c.h>
#include "snzip.h"

#define SNZ_MAGIC "SNZ"
#define SNZ_MAGIC_LEN 3
#define SNZ_FILE_VERSION 1

#define SNZ_DEFAULT_BLOCK_SIZE 16 /* (1 << 16) => 64 KiB */
#define SNZ_MAX_BLOCK_SIZE 27 /* (1 << 27) => 128 MiB */

#define VARINT_MAX 5

typedef struct {
  char magic[SNZ_MAGIC_LEN]; /* SNZ_MAGIC */
  char version;  /* SNZ_FILE_VERSION */
  unsigned char block_size; /* nth power of two. */
} snz_header_t;

static int snzip_compress(FILE *infp, FILE *outfp, size_t block_size)
{
  snz_header_t header;
  work_buffer_t wb;
  size_t uncompressed_length;
  int err = 1;
  int nshift;

  wb.c = NULL;
  wb.uc = NULL;

  if (block_size == 0) {
    block_size = 1ul << SNZ_DEFAULT_BLOCK_SIZE;
    nshift = SNZ_DEFAULT_BLOCK_SIZE;
  } else {
    if (block_size > (1ul << SNZ_MAX_BLOCK_SIZE)) {
      print_error("too large block size: %lu\n", block_size);
      goto cleanup;
    }

    for (nshift = 1; nshift <= SNZ_MAX_BLOCK_SIZE; nshift++) {
      if (1ul << nshift == block_size) {
        break;
      }
    }
    if (nshift == SNZ_MAX_BLOCK_SIZE) {
      print_error("The block size must be power of two\n");
      goto cleanup;
    }
  }

  /* write the file header */
  memcpy(header.magic, SNZ_MAGIC, SNZ_MAGIC_LEN);
  header.version = SNZ_FILE_VERSION;
  header.block_size = nshift;

  if (fwrite(&header, sizeof(header), 1, outfp) != 1) {
    print_error("Failed to write a file: %s\n", strerror(errno));
    goto cleanup;
  }

  /* write file body */
  work_buffer_init(&wb, block_size);
  while ((uncompressed_length = fread(wb.uc, 1, wb.uclen, infp)) > 0) {
    size_t compressed_length = wb.clen;

    trace("read %lu bytes.\n", (unsigned long)uncompressed_length);

    /* compress the block. */
    snappy_compress(wb.uc, uncompressed_length, wb.c, &compressed_length);
    trace("compressed_legnth is %lu.\n", (unsigned long)compressed_length);

    /* write the compressed length. */
    if (compressed_length < (1ul << 7)) {
      putc(compressed_length, outfp);
      trace("write 1 byte for compressed data length.\n");
    } else if (compressed_length < (1ul << 14)) {
      putc((compressed_length >> 0) | 0x80, outfp);
      putc((compressed_length >> 7), outfp);
      trace("write 2 bytes for compressed data length.\n");
    } else if (compressed_length < (1ul << 21)) {
      putc((compressed_length >> 0) | 0x80, outfp);
      putc((compressed_length >> 7) | 0x80, outfp);
      putc((compressed_length >> 14), outfp);
      trace("write 3 bytes for compressed data length.\n");
    } else if (compressed_length < (1ul << 28)) {
      putc((compressed_length >> 0) | 0x80, outfp);
      putc((compressed_length >> 7) | 0x80, outfp);
      putc((compressed_length >> 14)| 0x80, outfp);
      putc((compressed_length >> 21), outfp);
      trace("write 4 bytes for compressed data length.\n");
    } else {
      putc((compressed_length >> 0) | 0x80, outfp);
      putc((compressed_length >> 7) | 0x80, outfp);
      putc((compressed_length >> 14)| 0x80, outfp);
      putc((compressed_length >> 21)| 0x80, outfp);
      putc((compressed_length >> 28), outfp);
      trace("write 5 bytes for compressed data length.\n");
    }

    /* write the compressed data. */
    if (fwrite(wb.c, compressed_length, 1, outfp) != 1) {
      print_error("Failed to write a file: %s\n", strerror(errno));
      goto cleanup;
    }
    trace("write %ld bytes for compressed data.\n", (long)compressed_length);
  }
  if (!feof(infp)) {
    /* fread() failed. */
    print_error("Failed to read a file: %s\n", strerror(errno));
    goto cleanup;
  }
  putc('\0', outfp);
  trace("write 1 byte\n");
  err = 0;
 cleanup:
  work_buffer_free(&wb);
  return err;
}

static int snzip_uncompress(FILE *infp, FILE *outfp, int skip_magic)
{
  snz_header_t header;
  work_buffer_t wb;
  int err = 1;
  int outfd;

  wb.c = NULL;
  wb.uc = NULL;

  if (skip_magic) {
    header.block_size = snzip_format_block_size;
  } else {
    /* read header */
    if (fread(&header, sizeof(header), 1, infp) != 1) {
      print_error("Failed to read a file: %s\n", strerror(errno));
      goto cleanup;
    }

    /* check magic */
    if (memcmp(header.magic, SNZ_MAGIC, SNZ_MAGIC_LEN) != 0) {
      print_error("This is not a snz file.\n");
      goto cleanup;
    }
    /* check rest header */
    if (header.version != SNZ_FILE_VERSION) {
      print_error("Unknown snz version %d\n", header.version);
      goto cleanup;
    }
  }
  if (header.block_size > SNZ_MAX_BLOCK_SIZE) {
    print_error("Invalid block size %d (max %d)\n", header.block_size, SNZ_MAX_BLOCK_SIZE);
    goto cleanup;
  }

  /* Use a file descriptor 'outfd' instead of the stdio file pointer 'outfp'
   * to reduce the number of write system calls.
   */
  fflush(outfp);
  outfd = fileno(outfp);

  /* read body */
  work_buffer_init(&wb, (1 << header.block_size));
  for (;;) {
    /* read the compressed length in a block */
    size_t compressed_length = 0;
    size_t uncompressed_length = wb.uclen;
    int idx;

    for (idx = 0; idx < VARINT_MAX; idx++) {
      int chr = getc(infp);
      if (chr == -1) {
        print_error("Unexpected end of file.\n");
        goto cleanup;
      }
      compressed_length |= ((chr & 127) << (idx * 7));
      if ((chr & 128) == 0) {
        break;
      }
    }
    trace("read %d bytes (compressed_length = %ld)\n", idx + 1, (long)compressed_length);
    if (idx == VARINT_MAX) {
      print_error("Invalid format.\n");
      goto cleanup;
    }
    if (compressed_length == 0) {
      /* read all blocks */
      err = 0;
      goto cleanup;
    }
    if (compressed_length > wb.clen) {
      print_error("Invalid data: too long compressed length\n");
      goto cleanup;
    }

    /* read the compressed data */
    if (fread(wb.c, compressed_length, 1, infp) != 1) {
      if (feof(infp)) {
        print_error("Unexpected end of file\n");
      } else {
        print_error("Failed to read a file: %s\n", strerror(errno));
      }
      goto cleanup;
    }
    trace("read %ld bytes.\n", (long)(compressed_length));

    /* check the uncompressed length */
    err = snappy_uncompressed_length(wb.c, compressed_length, &uncompressed_length);
    if (err != 0) {
      print_error("Invalid data: GetUncompressedLength failed %d\n", err);
      goto cleanup;
    }
    err = 1;
    if (uncompressed_length > wb.uclen) {
      print_error("Invalid data: too long uncompressed length\n");
      goto cleanup;
    }

    /* uncompress and write */
    if (snappy_uncompress(wb.c, compressed_length, wb.uc, &uncompressed_length)) {
      print_error("Invalid data: RawUncompress failed\n");
      goto cleanup;
    }
    if (write_full(outfd, wb.uc, uncompressed_length) != uncompressed_length) {
      print_error("Failed to write a file: %s\n", strerror(errno));
      goto cleanup;
    }
    trace("write %ld bytes\n", (long)uncompressed_length);
  }
 cleanup:
  work_buffer_free(&wb);
  return err;
}

stream_format_t snzip_format = {
  "snzip",
  "https://github.com/kubo/snzip",
  "snz",
  snzip_compress,
  snzip_uncompress,
};
